library/multichoicerecyclerview/src/main/java/com/davidecirillo/multichoicerecyclerview/MultiChoiceAdapter.java

```diff
@@ -315,24 +315,12 @@ private void performAll(Action action) {
     }
 
     private void refreshDataSetInternal() {
-        if (mItemList.isEmpty()) {
-            for (int i = 0; i < getItemCount(); i++) {
-                mItemList.put(i, State.INACTIVE);
-            }
-        } else if (mItemList.size() > getItemCount()) { // We have new items, let's add only the new ones as inactive
-            for (int i = 0; i < getItemCount(); i++) {
-                if (mItemList.get(i) != null) {
-                    mItemList.put(i, State.INACTIVE);
-                }
-            }
-        } else {
-            // Keep just the number of item in the getItemCount remove all the others.
-            for (int i = 0; i < mItemList.size(); i++) {
-                if (i + 1 > getItemCount()) {
-                    mItemList.remove(i);
-                }
-            }
+        mItemList.clear();
+
+        for (int i = 0; i < getItemCount(); i++) {
+            mItemList.put(i, State.INACTIVE);
         }
```

Comment:
I'll need to think about what's best for fixing this and providing good api that fits all the needs. In the mean time if you want to creare a PR that follow our contribution guidelines, along with testing i'll be more than happy to review it. You can take as base this branch and change the logic inside the `refreshDataSetInternal`. 

I think at this point we should just add all the item with an UNSELECTED state as you did the first time, so it'll be just 4 line change to this PR + testing for merging this into dev.